### ISSUE TITLE
Proposal: Allow IP based policy enforcement on ingress and egress

### ISSUE THREAD

**Issue by tsenart** [cilium/cilium#164]
It's useful to allow policy enforcement on  packets going to external destinations not managed by Cilium. i.e. A legacy database not running in a Cilium enabled container or in a separate network segment.

**Comment by tgraf** [comment]
How would a useful configuration API look like. Our policy currently allows to specify a coverage which translates to the endpoints to which the configuration should be allowed to. We have an existing "allow" rule to list allowed consumers. Would it make sense to have a "allowed-destinations" option which is a []net.IPNet or something similar.

Example:

```
     "coverage": ["io.cilium.app=web"],
     "allow-destinations": ["10.1.1.0/24", "!10.1.1.1"],
```

If no `allow-destinations` is present, then the default remains all destinations are allowed. As soon as a single `allow-destinations` is present, the only the sum of all allowed - disallowed destinations is allowed. 

Makes sense?

**Comment by tsenart** [comment]
Sounds good. However, wondering what the current backwards compatibility guarantees are for the configuration API. Can we still break things to make them nicer at this point?

I'd change `allow` to `ingress` or `incoming` and the `allow-destinations` to `egress` or `outgoing` while keeping the current whitelist behaviour.

**Comment by jrajahalme** [comment]
Created a new branch (164-allow-IP-based-policy-enforcement), currently has code for bpf probe for the LPM trie map type.

**Comment by jrajahalme** [comment]
Branch 164-allow-IP-based-policy-enforcement now has working code for ingress CIDR rules for both IPv4 and IPv6. Egress side is still work in progress, but it would be good to get feedback at this point.

**Comment by jrajahalme** [comment]
Addressed comments from @tgraf and rebased branch to current master.
Added egress processing (last two patches). Haven't had a change to test the egress parts yet.

**Comment by tgraf** [comment]
Fixed by #886
